File: dart/blockchain/lib/wallet_manager.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:two_finance_blockchain/blockchain/keys/keys.dart';
import 'package:two_finance_blockchain/blockchain/transaction/transaction.dart';
import 'package:two_finance_blockchain/blockchain/utils/json.dart';

const int _walletVersion = 1;
const Duration _unlockDuration = Duration(minutes: 2);
const String _walletFileAssociatedData = 'wallet-manager-file:v1';
const String _privateKeyAssociatedDataPrefix = 'wallet-manager-private-key:';

enum WalletAction { exportPrivateKey, changePassword, deleteWallet, withdraw }

class KeysetKDFParams {
  final String alg;
  final int time;
  final int memoryKB;
  final int parallel;
  final int keyLen;
  final Uint8List salt;

  KeysetKDFParams({
    required this.alg,
    required this.time,
    required this.memoryKB,
    required this.parallel,
    required this.keyLen,
    required this.salt,
  });

  factory KeysetKDFParams.create() {
    return KeysetKDFParams(
      alg: 'argon2id',
      time: 3,
      memoryKB: 64 * 1024,
      parallel: 1,
      keyLen: 32,
      salt: _randomBytes(16),
    );
  }

  factory KeysetKDFParams.fromJson(Map<String, dynamic> json) {
    return KeysetKDFParams(
      alg: json['alg'] as String,
      time: json['time'] as int,
      memoryKB: json['memory_kb'] as int,
      parallel: json['parallel'] as int,
      keyLen: json['key_len'] as int,
      salt: _decodeBytes(json['salt']),
    );
  }

  Map<String, dynamic> toJson() => {
    'alg': alg,
    'time': time,
    'memory_kb': memoryKB,
    'parallel': parallel,
    'key_len': keyLen,
    'salt': base64Encode(salt),
  };
}

class LocalEncryptedWalletFile {
  final KeysetKDFParams kdf;
  final Uint8List cipher;

  LocalEncryptedWalletFile({required this.kdf, required this.cipher});

  factory LocalEncryptedWalletFile.fromJson(Map<String, dynamic> json) {
    return LocalEncryptedWalletFile(
      kdf: KeysetKDFParams.fromJson(json['kdf'] as Map<String, dynamic>),
      cipher: _decodeBytes(json['cipher']),
    );
  }

  Map<String, dynamic> toJson() => {
    'kdf': kdf.toJson(),
    'cipher': base64Encode(cipher),
  };
}

class WalletFile {
  final int version;
  final String owner;
  final Uint8List encryptedPrivateKey;
  final KeysetKDFParams privateKeyKdf;
  final DateTime createdAt;
  final DateTime updatedAt;

  WalletFile({
    required this.version,
    required this.owner,
    required this.encryptedPrivateKey,
    required this.privateKeyKdf,
    required this.createdAt,
    required this.updatedAt,
  });

  factory WalletFile.fromJson(Map<String, dynamic> json) {
    return WalletFile(
      version: json['version'] as int,
      owner: json['owner'] as String,
      encryptedPrivateKey: _decodeBytes(json['encrypted_private_key']),
      privateKeyKdf: KeysetKDFParams.fromJson(
        json['private_key_kdf'] as Map<String, dynamic>,
      ),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
    'version': version,
    'owner': owner,
    'encrypted_private_key': base64Encode(encryptedPrivateKey),
    'private_key_kdf': privateKeyKdf.toJson(),
    'created_at': createdAt.toUtc().toIso8601String(),
    'updated_at': updatedAt.toUtc().toIso8601String(),
  };
}

class SignTransactionInput {
  final int chainID;
  final String from;
  final String to;
  final String method;
  final JsonMessage data;
  final int version;
  final String uuid7;
  final AuthorizationEnvelope? authorization;

  SignTransactionInput({
    required this.chainID,
    required this.from,
    required this.to,
    required this.method,
    required this.data,
    required this.version,
    required this.uuid7,
    this.authorization,
  });
}

class PreparedTransaction {
  final int chainID;
  final String from;
  final String to;
  final String method;
  final JsonMessage data;
  final int version;
  final String uuid7;
  final String hash;
  final String signature;
  final AuthorizationEnvelope? authorization;

  PreparedTransaction({
    required this.chainID,
    required this.from,
    required this.to,
    required this.method,
    required this.data,
    required this.version,
    required this.uuid7,
    this.hash = '',
    this.signature = '',
    this.authorization,
  });

  factory PreparedTransaction.fromJson(Map<String, dynamic> json) {
    return PreparedTransaction(
      chainID: json['chain_id'] as int,
      from: json['from'] as String,
      to: json['to'] as String,
      method: json['method'] as String,
      data: Map<String, dynamic>.from(json['data'] as Map),
      version: json['version'] as int,
      uuid7: json['uuid7'] as String,
      hash: json['hash'] as String? ?? '',
      signature: json['signature'] as String? ?? '',
      authorization: json['authorization'] == null
          ? null
          : AuthorizationEnvelope.fromJson(
              Map<String, dynamic>.from(json['authorization'] as Map),
            ),
    );
  }

  Map<String, dynamic> toJson() => {
    'chain_id': chainID,
    'from': from,
    'to': to,
    'method': method,
    'data': data,
    'version': version,
    'uuid7': uuid7,
    'hash': hash,
    'signature': signature,
    if (authorization != null) 'authorization': authorization!.toJson(),
  };
}

class SignedTransaction extends PreparedTransaction {
  SignedTransaction({
    required super.chainID,
    required super.from,
    required super.to,
    required super.method,
    required super.data,
    required super.version,
    required super.uuid7,
    required super.hash,
    required super.signature,
    super.authorization,
  });

  factory SignedTransaction.fromTransaction(Transaction tx) {
    return SignedTransaction(
      chainID: tx.chainID,
      from: tx.from,
      to: tx.to,
      method: tx.method,
      data: tx.data,
      version: tx.version,
      uuid7: tx.uuid7,
      hash: tx.hash,
      signature: tx.signature,
      authorization: tx.authorization,
    );
  }

  factory SignedTransaction.fromJson(Map<String, dynamic> json) {
    return SignedTransaction(
      chainID: json['chain_id'] as int,
      from: json['from'] as String,
      to: json['to'] as String,
      method: json['method'] as String,
      data: Map<String, dynamic>.from(json['data'] as Map),
      version: json['version'] as int,
      uuid7: json['uuid7'] as String,
      hash: json['hash'] as String,
      signature: json['signature'] as String,
      authorization: json['authorization'] == null
          ? null
          : AuthorizationEnvelope.fromJson(
              Map<String, dynamic>.from(json['authorization'] as Map),
            ),
    );
  }

  Transaction toNetworkTransaction() {
    return Transaction(
      chainID: chainID,
      from: from,
      to: to,
      method: method,
      data: data,
      version: version,
      uuid7: uuid7,
      hash: hash,
      signature: signature,
      authorization: authorization,
    );
  }
}

abstract class IWalletManager {
  Future<void> importPrivateKey(Uint8List privateKey, String password);
  Future<void> lock();
  Future<void> unlockWithPassword(String password);
  bool isUnlocked();
  Future<void> changePassword(String currentPassword, String newPassword);
  String get ownerAddress;
  Future<Transaction> signTransaction(SignTransactionInput input);
  Future<SignedTransaction> signPreparedTransaction(PreparedTransaction input);
}

class WalletManager implements IWalletManager {
  final String filePath;
  String _owner = '';
  Uint8List? _privateKey;
  DateTime? _unlockedUntil;
  Timer? _lockTimer;

  WalletManager(this.filePath);

  @override
  String get ownerAddress => _owner;

  @override
  Future<void> importPrivateKey(Uint8List privateKey, String password) async {
    if (password.isEmpty) {
      throw ArgumentError('password is required');
    }
    if (privateKey.isEmpty) {
      throw ArgumentError('private key is required');
    }
    if (filePath.isEmpty) {
      throw StateError('wallet file path is required');
    }

    final privateKeyHex = utf8.decode(privateKey);
    final owner = await _publicKeyFromPrivateHex(privateKeyHex);
    KeyManager.validateEDDSAPublicKeyHex(owner);

    final privateKeyKdf = KeysetKDFParams.create();
    final encryptedPrivateKey = await _encryptWithPassword(
      privateKey,
      password,
      privateKeyKdf,
      utf8.encode('$_privateKeyAssociatedDataPrefix$owner'),
    );

    final now = DateTime.now().toUtc();
    final walletFile = WalletFile(
      version: _walletVersion,
      owner: owner,
      encryptedPrivateKey: encryptedPrivateKey,
      privateKeyKdf: privateKeyKdf,
      createdAt: now,
      updatedAt: now,
    );

    final payload = utf8.encode(jsonEncode(walletFile.toJson()));
    final localFile = await _encryptWalletPayload(payload, password);
    await _writeEncryptedWalletFile(localFile);

    _owner = owner;
    _clearBytes(privateKey);
    await lock();
  }

  @override
  Future<void> lock() async {
    _lockTimer?.cancel();
    _lockTimer = null;
    final privateKey = _privateKey;
    if (privateKey != null) {
      _clearBytes(privateKey);
    }
    _privateKey = null;
    _unlockedUntil = null;
  }

  @override
  Future<void> unlockWithPassword(String password) async {
    if (password.isEmpty) {
      throw ArgumentError('password is required');
    }
    if (filePath.isEmpty) {
      throw StateError('wallet file path is required');
    }

    final walletFile = await _readWalletFile(password);
    if (_owner.isNotEmpty && walletFile.owner != _owner) {
      throw StateError('wallet owner mismatch');
    }

    final privateKey = await _decryptWithPassword(
      walletFile.encryptedPrivateKey,
      password,
      walletFile.privateKeyKdf,
      utf8.encode('$_privateKeyAssociatedDataPrefix${walletFile.owner}'),
    );

    await lock();
    _owner = walletFile.owner;
    _privateKey = Uint8List.fromList(privateKey);
    _unlockedUntil = DateTime.now().add(_unlockDuration);
    _lockTimer = Timer(_unlockDuration, () {
      unawaited(lock());
    });
  }

  @override
  bool isUnlocked() {
    final privateKey = _privateKey;
    final unlockedUntil = _unlockedUntil;
    return privateKey != null &&
        privateKey.isNotEmpty &&
        unlockedUntil != null &&
        DateTime.now().isBefore(unlockedUntil);
  }

  @override
  Future<void> changePassword(
    String currentPassword,
    String newPassword,
  ) async {
    if (currentPassword.isEmpty) {
      throw ArgumentError('current password is required');
    }
    if (newPassword.isEmpty) {
      throw ArgumentError('new password is required');
    }
    if (currentPassword == newPassword) {
      throw ArgumentError(
        'new password must be different from current password',
      );
    }
    if (_owner.isEmpty) {
      throw StateError('owner is required');
    }

    final walletFile = await _readWalletFile(currentPassword);
    if (walletFile.owner != _owner) {
      throw StateError('wallet owner mismatch');
    }

    final privateKey = await _decryptWithPassword(
      walletFile.encryptedPrivateKey,
      currentPassword,
      walletFile.privateKeyKdf,
      utf8.encode('$_privateKeyAssociatedDataPrefix${walletFile.owner}'),
    );

    final newPrivateKeyKdf = KeysetKDFParams.create();
    final encryptedPrivateKey = await _encryptWithPassword(
      privateKey,
      newPassword,
      newPrivateKeyKdf,
      utf8.encode('$_privateKeyAssociatedDataPrefix${walletFile.owner}'),
    );

    final updatedWalletFile = WalletFile(
      version: walletFile.version,
      owner: walletFile.owner,
      encryptedPrivateKey: encryptedPrivateKey,
      privateKeyKdf: newPrivateKeyKdf,
      createdAt: walletFile.createdAt,
      updatedAt: DateTime.now().toUtc(),
    );

    final payload = utf8.encode(jsonEncode(updatedWalletFile.toJson()));
    final localFile = await _encryptWalletPayload(payload, newPassword);
    await _writeEncryptedWalletFile(localFile);
    _clearBytes(Uint8List.fromList(privateKey));
    await lock();
  }

  @override
  Future<Transaction> signTransaction(SignTransactionInput input) async {
    if (input.method.isEmpty) {
      throw ArgumentError('method is required');
    }
    if (!isUnlocked()) {
      throw StateError('wallet is locked');
    }

    final privateKey = Uint8List.fromList(_privateKey!);
    try {
      final tx = Transaction.create(
        chainID: input.chainID,
        from: input.from,
        to: input.to,
        method: input.method,
        data: input.data,
        version: input.version,
        uuid7: input.uuid7,
        authorization: input.authorization,
      );
      return signTransactionWithPrivateKey(privateKey, tx);
    } finally {
      _clearBytes(privateKey);
    }
  }

  @override
  Future<SignedTransaction> signPreparedTransaction(
    PreparedTransaction input,
  ) async {
    final signed = await signTransaction(
      SignTransactionInput(
        chainID: input.chainID,
        from: input.from,
        to: input.to,
        method: input.method,
        data: input.data,
        version: input.version,
        uuid7: input.uuid7,
        authorization: input.authorization,
      ),
    );
    return SignedTransaction.fromTransaction(signed);
  }

  Future<WalletFile> _readWalletFile(String password) async {
    final encryptedFile = File(filePath);
    if (!encryptedFile.existsSync()) {
      throw StateError('failed to read wallet file');
    }

    final localFile = LocalEncryptedWalletFile.fromJson(
      jsonDecode(await encryptedFile.readAsString()) as Map<String, dynamic>,
    );
    final payload = await _decryptWithPassword(
      localFile.cipher,
      password,
      localFile.kdf,
      utf8.encode(_walletFileAssociatedData),
    );
    final walletFile = WalletFile.fromJson(
      jsonDecode(utf8.decode(payload)) as Map<String, dynamic>,
    );

    if (walletFile.version != _walletVersion) {
      throw StateError('unsupported wallet version: ${walletFile.version}');
    }
    if (walletFile.owner.isEmpty) {
      throw StateError('wallet owner is required');
    }
    if (walletFile.encryptedPrivateKey.isEmpty) {
      throw StateError('encrypted private key is required');
    }
    return walletFile;
  }

  Future<LocalEncryptedWalletFile> _encryptWalletPayload(
    List<int> payload,
    String password,
  ) async {
    final kdf = KeysetKDFParams.create();
    final cipher = await _encryptWithPassword(
      payload,
      password,
      kdf,
      utf8.encode(_walletFileAssociatedData),
    );
    return LocalEncryptedWalletFile(kdf: kdf, cipher: cipher);
  }

  Future<void> _writeEncryptedWalletFile(
    LocalEncryptedWalletFile localFile,
  ) async {
    final walletFile = File(filePath);
    await walletFile.parent.create(recursive: true);
    await walletFile.writeAsString(jsonEncode(localFile.toJson()));
  }
}

Future<Transaction> signTransactionWithPrivateKey(
  Uint8List privateKey,
  Transaction tx,
) {
  return signTransaction(utf8.decode(privateKey), tx);
}

Future<KeyPair2Finance> generateEd25519KeyPairHex() {
  return KeyManager().generateKeyEd25519();
}

Future<String> _publicKeyFromPrivateHex(String privateKeyHex) async {
  final privateKeyBytes = KeyManager.hexToBytes(privateKeyHex);
  if (privateKeyBytes.length < 32) {
    throw ArgumentError('private key must be at least 32 bytes');
  }
  final keyPair = await Ed25519().newKeyPairFromSeed(
    privateKeyBytes.sublist(0, 32),
  );
  final publicKey = await keyPair.extractPublicKey();
  return KeyManager.bytesToHex(publicKey.bytes);
}

Future<Uint8List> _encryptWithPassword(
  List<int> data,
  String password,
  KeysetKDFParams params,
  List<int> associatedData,
) async {
  final key = await _derivePasswordKey(password, params);
  final box = await Xchacha20.poly1305Aead().encrypt(
    data,
    secretKey: key,
    aad: associatedData,
  );
  return Uint8List.fromList(box.concatenation());
}

Future<List<int>> _decryptWithPassword(
  List<int> cipher,
  String password,
  KeysetKDFParams params,
  List<int> associatedData,
) async {
  final key = await _derivePasswordKey(password, params);
  final box = SecretBox.fromConcatenation(
    cipher,
    nonceLength: Xchacha20.poly1305Aead().nonceLength,
    macLength: 16,
  );
  try {
    return await Xchacha20.poly1305Aead().decrypt(
      box,
      secretKey: key,
      aad: associatedData,
    );
  } on SecretBoxAuthenticationError {
    throw StateError('invalid password or corrupted keyset');
  }
}

Future<SecretKey> _derivePasswordKey(String password, KeysetKDFParams params) {
  if (password.isEmpty) {
    throw ArgumentError('password is required');
  }
  if (params.alg != 'argon2id') {
    throw ArgumentError('unsupported KDF algorithm: ${params.alg}');
  }
  if (params.salt.isEmpty) {
    throw ArgumentError('salt is required');
  }
  return Argon2id(
    parallelism: params.parallel,
    memory: params.memoryKB,
    iterations: params.time,
    hashLength: params.keyLen,
  ).deriveKey(secretKey: SecretKey(utf8.encode(password)), nonce: params.salt);
}

Uint8List _randomBytes(int length) {
  final random = Random.secure();
  return Uint8List.fromList(
    List<int>.generate(length, (_) => random.nextInt(256)),
  );
}

Uint8List _decodeBytes(dynamic value) {
  if (value is String) {
    return Uint8List.fromList(base64Decode(value));
  }
  if (value is List) {
    return Uint8List.fromList(value.cast<int>());
  }
  throw ArgumentError('invalid byte encoding');
}

void _clearBytes(Uint8List data) {
  for (var i = 0; i < data.length; i++) {
    data[i] = 0;
  }
}

File: go/wallet_manager/wallet_manager.go
package wallet_manager

import (
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"gitlab.com/2finance/2finance-network/blockchain/encryption/keys"
	"gitlab.com/2finance/2finance-network/blockchain/transaction"
	"gitlab.com/2finance/2finance-network/blockchain/utils"

	"github.com/2Finance-Labs/2finance-sdk-client/protocol"
)

const (
	walletVersion  = 1
	unlockDuration = 2 * time.Minute
)

type WalletAction string

const (
	ActionExportPrivateKey WalletAction = "export_private_key"
	ActionChangePassword   WalletAction = "change_password"
	ActionDeleteWallet     WalletAction = "delete_wallet"
	ActionWithdraw         WalletAction = "withdraw"
)

type WalletFile struct {
	Version             int               `json:"version"`
	Owner               string            `json:"owner"`
	EncryptedPrivateKey []byte            `json:"encrypted_private_key"`
	WrappedKeyset       WrappedTinkKeyset `json:"wrapped_keyset"`
	CreatedAt           time.Time         `json:"created_at"`
	UpdatedAt           time.Time         `json:"updated_at"`
}

type WalletManager struct {
	mu sync.RWMutex

	filePath string
	owner    string

	privateKey []byte

	unlockedUntil time.Time
	lockTimer     *time.Timer
}

type SignTransactionInput struct {
	ChainID       uint8
	From          string
	To            string
	Method        string
	Data          map[string]interface{}
	Version       uint8
	UUID7         string
	Authorization *transaction.AuthorizationEnvelope
}

type IWalletManager interface {
	ImportPrivateKey(privateKey []byte, password string) error

	Lock() error
	UnlockWithPassword(password string) error
	IsUnlocked() bool

	ChangePassword(currentPassword string, newPassword string) error

	OwnerAddress() string

	SignTransaction(input SignTransactionInput) (*transaction.Transaction, error)
	SignPreparedTransaction(input protocol.PreparedTransaction) (protocol.SignedTransaction, error)
}

func NewWalletManager(filePath string) IWalletManager {
	return &WalletManager{
		filePath: filePath,
	}
}

func (w *WalletManager) ImportPrivateKey(privateKey []byte, password string) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if password == "" {
		return errors.New("password is required")
	}

	if len(privateKey) == 0 {
		return fmt.Errorf("private key is required")
	}

	if w.filePath == "" {
		return fmt.Errorf("wallet file path is required")
	}

	privateKeyHex := string(privateKey)

	publicKey, err := keys.PublicKeyFromEd25519PrivateHex(privateKeyHex)
	if err != nil {
		return fmt.Errorf("failed to derive public key from private key: %w", err)
	}

	owner := keys.PublicKeyToHex(publicKey)
	if owner == "" {
		return fmt.Errorf("owner is required")
	}

	w.owner = owner

	encryptionKey := NewEncryption(w.owner)

	kh, err := encryptionKey.NewAEAD()
	if err != nil {
		return fmt.Errorf("failed to create wallet AEAD: %w", err)
	}

	encryptedPrivateKey, err := encryptionKey.EncryptPrivateKey(privateKey)
	if err != nil {
		return fmt.Errorf("failed to encrypt private key: %w", err)
	}

	wrappedKeyset, err := WrapTinkKeyset(kh, password)
	if err != nil {
		return fmt.Errorf("failed to wrap keyset: %w", err)
	}

	now := time.Now()

	walletFile := WalletFile{
		Version:             walletVersion,
		Owner:               w.owner,
		EncryptedPrivateKey: encryptedPrivateKey,
		WrappedKeyset:       wrappedKeyset,
		CreatedAt:           now,
		UpdatedAt:           now,
	}

	walletPayload, err := json.Marshal(walletFile)
	if err != nil {
		return fmt.Errorf("failed to marshal wallet file: %w", err)
	}
	defer clearBytes(walletPayload)

	encryptionFile := NewEncryptionFile(w.filePath)

	localEncryptedWalletFile, err := encryptionFile.Encrypt(walletPayload, password)
	if err != nil {
		return fmt.Errorf("failed to encrypt wallet file: %w", err)
	}

	if err := encryptionFile.Write(*localEncryptedWalletFile); err != nil {
		return fmt.Errorf("failed to write wallet file: %w", err)
	}

	clearBytes(privateKey)
	w.lockMemoryLocked()

	return nil
}

func (w *WalletManager) Lock() error {
	w.mu.Lock()
	defer w.mu.Unlock()

	w.lockMemoryLocked()

	return nil
}

func (w *WalletManager) UnlockWithPassword(password string) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if password == "" {
		return errors.New("password is required")
	}

	if w.filePath == "" {
		return fmt.Errorf("wallet file path is required")
	}

	encryptionFile := NewEncryptionFile(w.filePath)

	localEncryptedWalletFile, err := encryptionFile.Read()
	if err != nil {
		return fmt.Errorf("failed to read wallet file: %w", err)
	}

	walletPayload, err := encryptionFile.Decrypt(*localEncryptedWalletFile, password)
	if err != nil {
		return fmt.Errorf("failed to decrypt wallet file: %w", err)
	}
	defer clearBytes(walletPayload)

	var walletFile WalletFile
	if err := json.Unmarshal(walletPayload, &walletFile); err != nil {
		return fmt.Errorf("failed to unmarshal wallet file: %w", err)
	}

	if walletFile.Version != walletVersion {
		return fmt.Errorf("unsupported wallet version: %d", walletFile.Version)
	}

	if walletFile.Owner == "" {
		return fmt.Errorf("wallet owner is required")
	}

	if len(walletFile.EncryptedPrivateKey) == 0 {
		return fmt.Errorf("encrypted private key is required")
	}

	if w.owner == "" {
		w.owner = walletFile.Owner
	}

	if walletFile.Owner != w.owner {
		return fmt.Errorf("wallet owner mismatch")
	}

	kh, err := UnwrapTinkKeyset(walletFile.WrappedKeyset, password)
	if err != nil {
		return fmt.Errorf("failed to unwrap keyset: %w", err)
	}

	encryptionKey := NewEncryption(walletFile.Owner)

	if err := encryptionKey.LoadAEAD(kh); err != nil {
		return fmt.Errorf("failed to load wallet AEAD: %w", err)
	}

	privateKey, err := encryptionKey.DecryptPrivateKey(walletFile.EncryptedPrivateKey)
	if err != nil {
		return fmt.Errorf("failed to decrypt private key: %w", err)
	}

	w.lockMemoryLocked()

	w.privateKey = privateKey
	w.owner = walletFile.Owner
	w.unlockedUntil = time.Now().Add(unlockDuration)

	w.lockTimer = time.AfterFunc(unlockDuration, func() {
		_ = w.Lock()
	})

	return nil
}

func (w *WalletManager) IsUnlocked() bool {
	w.mu.RLock()
	defer w.mu.RUnlock()

	return w.isUnlockedLocked()
}

func (w *WalletManager) ChangePassword(currentPassword string, newPassword string) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if currentPassword == "" {
		return errors.New("current password is required")
	}

	if newPassword == "" {
		return errors.New("new password is required")
	}

	if currentPassword == newPassword {
		return errors.New("new password must be different from current password")
	}

	if w.owner == "" {
		return fmt.Errorf("owner is required")
	}

	if w.filePath == "" {
		return fmt.Errorf("wallet file path is required")
	}

	encryptionFile := NewEncryptionFile(w.filePath)

	localEncryptedWalletFile, err := encryptionFile.Read()
	if err != nil {
		return fmt.Errorf("failed to read wallet file: %w", err)
	}

	walletPayload, err := encryptionFile.Decrypt(*localEncryptedWalletFile, currentPassword)
	if err != nil {
		return fmt.Errorf("failed to decrypt wallet file with current password: %w", err)
	}
	defer clearBytes(walletPayload)

	var walletFile WalletFile
	if err := json.Unmarshal(walletPayload, &walletFile); err != nil {
		return fmt.Errorf("failed to unmarshal wallet file: %w", err)
	}

	if walletFile.Version != walletVersion {
		return fmt.Errorf("unsupported wallet version: %d", walletFile.Version)
	}

	if walletFile.Owner != w.owner {
		return fmt.Errorf("wallet owner mismatch")
	}

	kh, err := UnwrapTinkKeyset(walletFile.WrappedKeyset, currentPassword)
	if err != nil {
		return fmt.Errorf("failed to unwrap keyset with current password: %w", err)
	}

	newWrappedKeyset, err := WrapTinkKeyset(kh, newPassword)
	if err != nil {
		return fmt.Errorf("failed to wrap keyset with new password: %w", err)
	}

	walletFile.WrappedKeyset = newWrappedKeyset
	walletFile.UpdatedAt = time.Now()

	updatedWalletPayload, err := json.Marshal(walletFile)
	if err != nil {
		return fmt.Errorf("failed to marshal updated wallet file: %w", err)
	}
	defer clearBytes(updatedWalletPayload)

	newLocalEncryptedWalletFile, err := encryptionFile.Encrypt(updatedWalletPayload, newPassword)
	if err != nil {
		return fmt.Errorf("failed to encrypt wallet file with new password: %w", err)
	}

	if err := encryptionFile.Write(*newLocalEncryptedWalletFile); err != nil {
		return fmt.Errorf("failed to write rotated wallet file: %w", err)
	}

	w.lockMemoryLocked()

	return nil
}

func (w *WalletManager) OwnerAddress() string {
	w.mu.RLock()
	defer w.mu.RUnlock()

	return w.owner
}

func (w *WalletManager) SignTransaction(input SignTransactionInput) (*transaction.Transaction, error) {
	if input.Method == "" {
		return nil, errors.New("method is required")
	}
	if len(input.Data) == 0 {
		return nil, errors.New("data cannot be empty")
	}

	w.mu.RLock()

	if !w.isUnlockedLocked() {
		w.mu.RUnlock()
		return nil, errors.New("wallet is locked")
	}

	privateKey := cloneBytes(w.privateKey)

	w.mu.RUnlock()
	defer clearBytes(privateKey)

	return w.signTransactionWithPrivateKey(
		privateKey,
		input.ChainID,
		input.From,
		input.To,
		input.Method,
		input.Data,
		input.Version,
		input.UUID7,
		input.Authorization,
	)
}

func (w *WalletManager) SignPreparedTransaction(input protocol.PreparedTransaction) (protocol.SignedTransaction, error) {
	signed, err := w.SignTransaction(SignTransactionInput{
		ChainID:       input.ChainID,
		From:          input.From,
		To:            input.To,
		Method:        input.Method,
		Data:          input.Data,
		Version:       input.Version,
		UUID7:         input.UUID7,
		Authorization: input.Authorization,
	})
	if err != nil {
		return protocol.SignedTransaction{}, err
	}

	return protocol.SignedTransactionFromNetwork(signed)
}

func GenerateEd25519KeyPairHex() (string, string, error) {
	publicKey, privateKey, err := keys.GenerateEd25519KeyPair()
	if err != nil {
		return "", "", fmt.Errorf("failed to generate public key: %w", err)
	}

	privateKeyHex := keys.PrivateKeyToHex(privateKey)
	publicKeyHex := keys.PublicKeyToHex(publicKey)

	return publicKeyHex, privateKeyHex, nil
}

func (w *WalletManager) exportPrivateKey(password string) ([]byte, error) {
	return w.privateKeyForAction(ActionExportPrivateKey, password)
}

func (w *WalletManager) privateKeyForAction(action WalletAction, password string) ([]byte, error) {
	if requiresPassword(action) {
		if password == "" {
			return nil, errors.New("password is required")
		}

		if err := w.UnlockWithPassword(password); err != nil {
			return nil, err
		}
	}

	if !w.IsUnlocked() {
		if password == "" {
			return nil, errors.New("wallet is locked")
		}

		if err := w.UnlockWithPassword(password); err != nil {
			return nil, err
		}
	}

	w.mu.RLock()
	defer w.mu.RUnlock()

	return cloneBytes(w.privateKey), nil
}

func requiresPassword(action WalletAction) bool {
	switch action {
	case ActionExportPrivateKey,
		ActionChangePassword,
		ActionDeleteWallet,
		ActionWithdraw:
		return true
	default:
		return false
	}
}

func (w *WalletManager) isUnlockedLocked() bool {
	return len(w.privateKey) > 0 && time.Now().Before(w.unlockedUntil)
}

func (w *WalletManager) lockMemoryLocked() {
	if w.lockTimer != nil {
		w.lockTimer.Stop()
		w.lockTimer = nil
	}

	if len(w.privateKey) > 0 {
		clearBytes(w.privateKey)
	}

	w.privateKey = nil
	w.unlockedUntil = time.Time{}
}

func (w *WalletManager) signTransactionWithPrivateKey(
	privateKey []byte,
	chainID uint8,
	from string,
	to string,
	method string,
	data utils.JSONB,
	version uint8,
	uuid7 string,
	authorization *transaction.AuthorizationEnvelope,
) (*transaction.Transaction, error) {
	if len(privateKey) == 0 {
		return nil, errors.New("private key is required")
	}

	dataRawMessage, err := utils.MapToRawMessage(data)
	if err != nil {
		return nil, fmt.Errorf("failed to convert data to RawMessage: %w", err)
	}

	newTx := transaction.NewTransaction(
		chainID,
		from,
		to,
		method,
		dataRawMessage,
		version,
		uuid7,
	)

	tx := newTx.Get()
	tx.Authorization = authorization

	signedTx, err := transaction.SignTransactionHexKey(string(privateKey), tx)
	if err != nil {
		return nil, fmt.Errorf("failed to sign transaction: %w", err)
	}

	return signedTx, nil
}

func (w *WalletManager) loadPrivateKeyOnce(password string) ([]byte, error) {
	w.mu.RLock()
	filePath := w.filePath
	owner := w.owner
	w.mu.RUnlock()

	if password == "" {
		return nil, errors.New("password is required")
	}

	if filePath == "" {
		return nil, fmt.Errorf("wallet file path is required")
	}

	encryptionFile := NewEncryptionFile(filePath)

	localEncryptedWalletFile, err := encryptionFile.Read()
	if err != nil {
		return nil, fmt.Errorf("failed to read wallet file: %w", err)
	}

	walletPayload, err := encryptionFile.Decrypt(*localEncryptedWalletFile, password)
	if err != nil {
		return nil, fmt.Errorf("failed to decrypt wallet file: %w", err)
	}
	defer clearBytes(walletPayload)

	var walletFile WalletFile
	if err := json.Unmarshal(walletPayload, &walletFile); err != nil {
		return nil, fmt.Errorf("failed to unmarshal wallet file: %w", err)
	}

	if walletFile.Version != walletVersion {
		return nil, fmt.Errorf("unsupported wallet version: %d", walletFile.Version)
	}

	if walletFile.Owner == "" {
		return nil, fmt.Errorf("wallet owner is required")
	}

	if len(walletFile.EncryptedPrivateKey) == 0 {
		return nil, fmt.Errorf("encrypted private key is required")
	}

	if owner != "" && walletFile.Owner != owner {
		return nil, fmt.Errorf("wallet owner mismatch")
	}

	kh, err := UnwrapTinkKeyset(walletFile.WrappedKeyset, password)
	if err != nil {
		return nil, fmt.Errorf("failed to unwrap keyset: %w", err)
	}

	encryptionKey := NewEncryption(walletFile.Owner)

	if err := encryptionKey.LoadAEAD(kh); err != nil {
		return nil, fmt.Errorf("failed to load wallet AEAD: %w", err)
	}

	privateKey, err := encryptionKey.DecryptPrivateKey(walletFile.EncryptedPrivateKey)
	if err != nil {
		return nil, fmt.Errorf("failed to decrypt private key: %w", err)
	}

	return privateKey, nil
}
